ceno_zkvm/src/scheme/utils.rs

```diff
@@ -233,118 +239,248 @@ pub(crate) fn infer_tower_product_witness<E: ExtensionField>(
     wit_layers
 }
 
+fn try_recycle_arcpoly<E: ExtensionField>(
+    poly: Cow<ArcMultilinearExtension<'_, E>>,
+    pool_e: &mut SimpleVecPool<Vec<E>>,
+    pool_b: &mut SimpleVecPool<Vec<E::BaseField>>,
+    pool_expected_size_vec: usize,
+) {
+    fn downcast_arc<E: ExtensionField>(
+        arc: ArcMultilinearExtension<'_, E>,
+    ) -> DenseMultilinearExtension<E> {
+        unsafe {
+            // get the raw pointer from the Arc
+            assert_eq!(Arc::strong_count(&arc), 1);
```

Comment:
This is to
- follow `ArcMultilinearExtension<'_, E>` type across the whole interface
- however, there are some intermediate `ArcMultilinearExtension<'_, E>` which is hold some intermediate vector from vector pool, and we assure it can be return. However, I got `double free or corruption` error